src/analyses/base.ml

```diff
@@ -321,9 +321,26 @@ struct
     | Address p, Int n
     | Int n, Address p when op=Eq || op=Ne ->
       let ik = Cilfacade.get_ikind t in
-      Int (match ID.to_bool n, AD.to_bool p with
-          | Some a, Some b -> ID.of_bool ik (op=Eq && a=b || op=Ne && a<>b)
-          | _ -> bool_top ik)
+      let res =
+        if AD.is_null p then
+          match (ID.to_bool n) with
+          | Some true ->
+            (* n is definitely not 0, p is NULL *)
+            ID.of_bool ik (op = Ne)
+          | Some false ->
+            (* n is zero, p is NULL*)
+            ID.of_bool ik (op = Eq)
+          | _ -> bool_top ik
+        else if AD.is_not_null p then
+          match (ID.to_int n) with
+          | Some i when Z.equal i Z.zero ->
+            (* n is zero, p is not NULL *)
+            ID.of_bool ik (op = Ne)
+          | _ -> bool_top ik
```

Comment:
Why does the `is_not_null` case use `ID.to_int` and `Z.equal` to check the same thing that `is_null` case does with `ID.to_bool`?

Also, `is_null` and `is_not_null` just give the two cases of `AD.to_bool` but with duplicate checking, no?
Seems to me that the original match over both `ID.to_bool` and `AD.to_bool` would still be a more concise way to do this double matching. The problem is just that it should exclude the case where both are non-null, where we know nothing about their equality.

Reply:
> Why does the is_not_null case use ID.to_int and Z.equal to check the same thing that is_null case does with ID.to_bool?

I think like this the intent is more clearly visible.


> the original match over both `ID.to_boo`l and `AD.to_bool` would still be more concise

More concise probably yes but I think the current version is much easier to read and understand and does not require reasoning about what `AD.to_bool` does. The fact that this bug was hidden for so long I think is a good argument to be as literal here as possible.

Reply:
Matching `ID.equal_to 0` would then make the intent the most clear.